pynumaflow/function/multiproc_server.py

```diff
@@ -0,0 +1,96 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import contextlib
+import os
+from concurrent import futures
+import datetime
+import logging
+import multiprocessing
+import socket
+import time
+import grpc
+
+from pynumaflow._constants import MULTIPROC_FUNCTION_SOCK_PATH, MULTIPROC_FUNCTION_SOCK_ADDR
+from pynumaflow.function.proto import udfunction_pb2_grpc
+
+from pynumaflow import setup_logging
+
+_LOGGER = setup_logging(__name__)
+if os.getenv("PYTHONDEBUG"):
+    _LOGGER.setLevel(logging.DEBUG)
+
+_ONE_DAY = datetime.timedelta(days=1)
+
+
+class MultiProcServer:
+    """
+    Provides a multiprocessing implementation of the grpc server.
+    They use a given TCP socket with SOCK_REUSE to allow all servers,
+    to listen on the same port. We start servers equal to the CPU count of the
+    system.
+    """
+
+    def __init__(self, udf_service, server_options):
+        self.udf_service = udf_service
+        self.sock_path = MULTIPROC_FUNCTION_SOCK_PATH
+        self._PROCESS_COUNT = multiprocessing.cpu_count()
+        self._THREAD_CONCURRENCY = self._PROCESS_COUNT
+        self.server_options = server_options
+
+    def _wait_forever(self, server):
+        try:
+            while True:
+                time.sleep(_ONE_DAY.total_seconds())
+        except KeyboardInterrupt:
+            server.stop(None)
+
+    def _run_server(self, bind_address):
+        """Start a server in a subprocess."""
+        _LOGGER.info("Starting new server.")
+        options = [("grpc.so_reuseport", 1), ("grpc.so_reuseaddr", 1)]
+        for x in options:
+            self.server_options.append(x)
+        server = grpc.server(
+            futures.ThreadPoolExecutor(
```

Comment:
what is threadpool doing here?

Reply:
That's how we initialize grpc servers in pynumaflow for other cases as well. Need to check if they actually make a difference in performance or not.